File: packages/types/src/contract.ts
/**
 * Copyright (c) 2023-present Plane Software, Inc. and contributors
 * SPDX-License-Identifier: AGPL-3.0-only
 * See the LICENSE file for details.
 */

// Mirrors apps/api/plane/db/models/contract.py (crm-new aligned schema)

export type TContractProcessingStatus = "PENDING" | "PROCESSING" | "COMPLETED" | "ERROR";
export type TContractStatus = "VIGENTE" | "FINALIZADO" | "NO_ESPECIFICADO";
export type TContractType = "ARRENDAMIENTOS" | "ALQUILERES" | "VEHICULOS" | "SERVICIOS" | "ARTISTAS";
export type TYesNoUnspecified = "SI" | "NO" | "NO_ESPECIFICADO";

export type TContract = {
  id: string;
  workspace_id: string;
  file_asset_id: string | null;
  thumbnail_asset_id: string | null;
  file_name: string | null;
  processing_status: TContractProcessingStatus;
  proposed_data: Record<string, unknown> | null;
  ai_model_used: string | null;
  processed_at: string | null;
  text_extracted_at: string | null;
  // AI-extracted, user-editable fields
  titulo: string | null;
  resumen_general: string | null;
  nombre_grupo: string | null;
  /** Comma-separated plain text (crm-new String @db.Text) */
  artistas: string | null;
  testigos: string | null;
  involucrados: string | null;
  es_notariado: boolean | null;
  fecha_inicio: string | null;
  fecha_fin: string | null;
  es_posible_expandirlo: TYesNoUnspecified;
  tiempo_extension_posible: string | null;
  expansion_time_description: string | null;
  fecha_fin_efectiva: string | null;
  estatus_contrato: TContractStatus;
  tipo_contrato: TContractType | null;
  periodo_coleccion: TYesNoUnspecified | null;
  collection_period_description: string | null;
  collection_period_duration: string | null;
  periodo_retencion: TYesNoUnspecified | null;
  retention_period_description: string | null;
  retention_period_duration: string | null;
  created_at: string;
  updated_at: string;
};

export type TContractUpdatePayload = Partial<
  Pick<
    TContract,
    | "titulo"
    | "resumen_general"
    | "nombre_grupo"
    | "artistas"
    | "testigos"
    | "involucrados"
    | "es_notariado"
    | "fecha_inicio"
    | "fecha_fin"
    | "es_posible_expandirlo"
    | "tiempo_extension_posible"
    | "expansion_time_description"
    | "fecha_fin_efectiva"
    | "estatus_contrato"
    | "tipo_contrato"
    | "periodo_coleccion"
    | "collection_period_description"
    | "collection_period_duration"
    | "periodo_retencion"
    | "retention_period_description"
    | "retention_period_duration"
  >
>;

export type TContractJobStatus = "QUEUED" | "RUNNING" | "COMPLETED" | "FAILED";
export type TContractJobTaskType = "EXTRACT_FULL" | "RETRY_PARTIAL" | "REANALYZE" | "QUERY";

export type TContractJob = {
  id: string;
  workspace_id: string;
  contract_id: string | null;
  initiated_by_id: string | null;
  task_type: TContractJobTaskType;
  status: TContractJobStatus;
  progress: number;
  current_stage: string | null;
  workflow_instance_id: string | null;
  error: { message?: string; stage?: string } | null;
  metadata: Record<string, unknown> | null;
  started_at: string | null;
  finished_at: string | null;
  created_at: string;
  updated_at: string;
};

export type TContractQueryMatch = {
  contract_id: string;
  title: string;
  artists: string | null;
  start_date: string | null;
  end_date: string | null;
  final_end_date: string | null;
  reason: string;
};

export type TContractQuery = {
  id: string;
  workspace_id: string;
  user_id: string;
  query: string;
  status: TContractJobStatus;
  result: { summary?: string; matches?: TContractQueryMatch[]; scanned_count?: number } | null;
  emailed_at: string | null;
  job_id: string | null;
  created_at: string;
};

export type TContractRetryOptions = {
  extract_text?: boolean;
  generate_embeddings?: boolean;
  ai_analysis?: boolean;
  extract_thumbnail?: boolean;
};

export type TContractFilters = {
  search?: string;
  person?: string;
  artist?: string;
  year?: string;
  /** OR'd within the group (multi-value) */
  estatus?: TContractStatus[];
  tipo?: TContractType[];
  processing_status?: TContractProcessingStatus[];
  fecha_fin_efectiva_after?: string;
  fecha_fin_efectiva_before?: string;
};

File: apps/api/plane/app/serializers/contract.py
# Copyright (c) 2023-present Plane Software, Inc. and contributors
# SPDX-License-Identifier: AGPL-3.0-only
# See the LICENSE file for details.

# Third party imports
from rest_framework import serializers

# Module imports
from plane.db.models import Contract, ContractChat, ContractChatMessage, ContractProcessingJob, ContractQuery

from .base import BaseSerializer

# AI-extracted fields the user can edit by hand (mirrors the crm-new schema)
CONTRACT_EDITABLE_FIELDS = [
    "titulo",
    "resumen_general",
    "nombre_grupo",
    "artistas",
    "testigos",
    "involucrados",
    "es_notariado",
    "fecha_inicio",
    "fecha_fin",
    "es_posible_expandirlo",
    "tiempo_extension_posible",
    "expansion_time_description",
    "fecha_fin_efectiva",
    "estatus_contrato",
    "tipo_contrato",
    "periodo_coleccion",
    "collection_period_description",
    "collection_period_duration",
    "periodo_retencion",
    "retention_period_description",
    "retention_period_duration",
]


class ContractSerializer(BaseSerializer):
    file_name = serializers.SerializerMethodField()

    class Meta:
        model = Contract
        fields = [
            "id",
            "workspace_id",
            "file_asset_id",
            "thumbnail_asset_id",
            "file_name",
            "processing_status",
            "proposed_data",
            "ai_model_used",
            "processed_at",
            "text_extracted_at",
            *CONTRACT_EDITABLE_FIELDS,
            "created_at",
            "updated_at",
        ]
        read_only_fields = fields

    def get_file_name(self, obj):
        return (obj.file_asset.attributes or {}).get("name") if obj.file_asset_id else None


class ContractUpdateSerializer(BaseSerializer):
    class Meta:
        model = Contract
        fields = CONTRACT_EDITABLE_FIELDS


class ContractProcessingJobSerializer(BaseSerializer):
    class Meta:
        model = ContractProcessingJob
        fields = [
            "id",
            "workspace_id",
            "contract_id",
            "initiated_by_id",
            "task_type",
            "status",
            "progress",
            "current_stage",
            "workflow_instance_id",
            "error",
            "metadata",
            "started_at",
            "finished_at",
            "created_at",
            "updated_at",
        ]
        read_only_fields = fields


class ContractChatSerializer(BaseSerializer):
    class Meta:
        model = ContractChat
        fields = [
            "id",
            "workspace_id",
            "user_id",
            "title",
            "mode",
            "contract_id",
            "created_at",
            "updated_at",
        ]
        read_only_fields = fields


class ContractChatMessageSerializer(BaseSerializer):
    class Meta:
        model = ContractChatMessage
        fields = ["id", "chat_id", "role", "content", "sources", "created_at"]
        read_only_fields = fields


class ContractQuerySerializer(BaseSerializer):
    class Meta:
        model = ContractQuery
        fields = [
            "id",
            "workspace_id",
            "user_id",
            "query",
            "status",
            "result",
            "emailed_at",
            "job_id",
            "created_at",
        ]
        read_only_fields = fields
